Adds IdmControlledMaliputRailcar to AutmotiveSimulator.

liangfok opened the pull request.

liangfok commented:
This PR is blocked on #5755.

liangfok commented:
+@jadecastro for feature review please since this is the `MaliputRailcar` version of #5288. Thanks!

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-:-KhUTxY7UP5Rn89VzBVi:b2y0rot)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Looks good. A few comments below. PTAL.

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-KhcF_4nuHFjKxl9H2sf:-KhcF_4nuHFjKxl9H2sg:buj1gds) ([raw file](https://github.com/robotlocomotion/drake/blob/9e57da35e16e10f3091ede342a07c32f43ae794d/drake/automotive/automotive_demo.cc#L27)):*
> ```cpp
>              "option is currently only applied when the road network is a flat "
>              "plane or a dragway.");
> DEFINE_int32(num_maliput_railcar, 0, "Number of MaliputRailcar vehicles. This "
> ```

BTW, please state here that this parameter sets the number of `MaliputRailcars` that are moving at a fixed speed.  Please also state that these are added after `num_trajectory_car` vehicles have been added. 

---

*[drake/automotive/automotive_demo.cc, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-KhcFG_uASU83kbCDTnr:-KhcFG_uASU83kbCDTns:be5sfsa) ([raw file](https://github.com/robotlocomotion/drake/blob/9e57da35e16e10f3091ede342a07c32f43ae794d/drake/automotive/automotive_demo.cc#L32)):*
> ```cpp
> DEFINE_int32(num_idm_controlled_maliput_railcar, 0, "Number of IDM-controlled "
>              "MaliputRailcar vehicles. This option is currently only applied "
>              "when the road network is a dragway.");
> ```

BTW, please specify that the IDM-controlled MaliputRailcars are added after `num_maliput_railcar` vehicles have been added.

---

*[drake/automotive/automotive_demo.cc, line 108 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-Khc7BcfmEKHFAlb_T8E:-Khc7BcfmEKHFAlb_T8F:bsriyhh) ([raw file](https://github.com/robotlocomotion/drake/blob/9e57da35e16e10f3091ede342a07c32f43ae794d/drake/automotive/automotive_demo.cc#L108)):*
> ```cpp
>     const int row = i / FLAGS_num_dragway_lanes;
>     // 5 m ensures a gap between consecutive rows of Prius vehicles. It was
>     // emperically chosen.
> ```

BTW `emperically` -> `empirically`

---

*[drake/automotive/automotive_demo.cc, line 109 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-KhcAqlb_oS2Qhw2rHBS:-KhcAqlchy6K6aeAWQOo:b4awm05) ([raw file](https://github.com/robotlocomotion/drake/blob/9e57da35e16e10f3091ede342a07c32f43ae794d/drake/automotive/automotive_demo.cc#L109)):*
> ```cpp
>     // 5 m ensures a gap between consecutive rows of Prius vehicles. It was
>     // emperically chosen.
>     state.set_s(5 * row);
> ```

BTW This is okay, but I would really like to have the ability to specify initial conditions `s` and `speed` as inputs, or perhaps with another option flag to create uniform random inputs within the upper/lower bounds for each parameter. Does any design exist yet for doing that? If not, it's fine to leave as is and come back to that later. 

---

*[drake/automotive/automotive_demo.cc, line 176 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-KhcFqqqc71c48q_kYO2:-KhcFqqqc71c48q_kYO3:bvfpmut) ([raw file](https://github.com/robotlocomotion/drake/blob/9e57da35e16e10f3091ede342a07c32f43ae794d/drake/automotive/automotive_demo.cc#L176)):*
> ```cpp
>                                        std::get<2>(params));
>     }
>     AddMaliputRailcar(FLAGS_num_maliput_railcar, false /* IDM controlled */,
> ```

BTW if someone has values for both `num_maliput_railcar` and `num_idm_controlled_maliput_railcar`, the vanilla `MaliputRailcars` are added on top of their IDM-controlled counterparts, which shouldn't happen. I think it makes most sense for the IDM-controlled railcars to be added first, then any constant-speed `MaliputRailcars` to be populated in front of them.

---

*[drake/automotive/automotive_simulator.h, line 139 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-KhcDl3XNzfcWAXgHhi9:-KhcDl3XNzfcWAXgHhiA:bi1txcc) ([raw file](https://github.com/robotlocomotion/drake/blob/9e57da35e16e10f3091ede342a07c32f43ae794d/drake/automotive/automotive_simulator.h#L139)):*
> ```objc
>   /// a std::runtime_error will be thrown.
>   ///
>   /// @param params The MaliputRailcar's parameters.
> ```

BTW perhaps you want to state that the field is optional and default parameters are selected if none are supplied here. 

Ditto with `MaliputRailcarState` and the input parameters for `AddPriusMaliputRailcar`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-:-KhcJfmBnXlb_zpdyCgE:b51b2ap)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@jadecastro, thanks for the review, I addressed your comments. PTAL.

---

Review status: 1 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-KhcF_4nuHFjKxl9H2sf:-KhcLkhPZMUX4oVOe_Km:b-z6fb4u) ([raw file](https://github.com/robotlocomotion/drake/blob/9e57da35e16e10f3091ede342a07c32f43ae794d/drake/automotive/automotive_demo.cc#L27)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, please state here that this parameter sets the number of `MaliputRailcars` that are moving at a fixed speed.  Please also state that these are added after `num_trajectory_car` vehicles have been added. 
</blockquote></details>

Done, though the fact that the Maliput railcars are added after the trajectory cars seems like an implementation detail that need not be conveyed to the user.

---

*[drake/automotive/automotive_demo.cc, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-KhcFG_uASU83kbCDTnr:-KhcLsnIcZLf5GKjIJFx:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9e57da35e16e10f3091ede342a07c32f43ae794d/drake/automotive/automotive_demo.cc#L32)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, please specify that the IDM-controlled MaliputRailcars are added after `num_maliput_railcar` vehicles have been added.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 108 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-Khc7BcfmEKHFAlb_T8E:-KhcM0uUoXYOTZar-7OQ:boyexbg) ([raw file](https://github.com/robotlocomotion/drake/blob/9e57da35e16e10f3091ede342a07c32f43ae794d/drake/automotive/automotive_demo.cc#L108)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW `emperically` -> `empirically`
</blockquote></details>

Done. Thanks.

---

*[drake/automotive/automotive_demo.cc, line 109 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-KhcAqlb_oS2Qhw2rHBS:-KhcM3KOTZGR9y5olyox:b-6814me) ([raw file](https://github.com/robotlocomotion/drake/blob/9e57da35e16e10f3091ede342a07c32f43ae794d/drake/automotive/automotive_demo.cc#L109)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW This is okay, but I would really like to have the ability to specify initial conditions `s` and `speed` as inputs, or perhaps with another option flag to create uniform random inputs within the upper/lower bounds for each parameter. Does any design exist yet for doing that? If not, it's fine to leave as is and come back to that later. 
</blockquote></details>

I agree, though no design for this exists yet. #5250 is a draft PR that uses Protocol Buffers to allow these types of parameters to be specified. Maybe we can discuss there?

---

*[drake/automotive/automotive_demo.cc, line 176 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-KhcFqqqc71c48q_kYO2:-KhcRb8jOTyl1lwFmo8e:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9e57da35e16e10f3091ede342a07c32f43ae794d/drake/automotive/automotive_demo.cc#L176)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW if someone has values for both `num_maliput_railcar` and `num_idm_controlled_maliput_railcar`, the vanilla `MaliputRailcars` are added on top of their IDM-controlled counterparts, which shouldn't happen. I think it makes most sense for the IDM-controlled railcars to be added first, then any constant-speed `MaliputRailcars` to be populated in front of them.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_simulator.h, line 139 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-KhcDl3XNzfcWAXgHhi9:-KhcS3wW6diusLGuzTjT:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9e57da35e16e10f3091ede342a07c32f43ae794d/drake/automotive/automotive_simulator.h#L139)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW perhaps you want to state that the field is optional and default parameters are selected if none are supplied here. 

Ditto with `MaliputRailcarState` and the input parameters for `AddPriusMaliputRailcar`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-:-KhcS5-_jG1UKJf1zEsC:b-tsqeui)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 3 files at r2.
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/automotive_demo.cc, line 109 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-KhcAqlb_oS2Qhw2rHBS:-KhdymY5j-_GQniWVIHw:bck4a05) ([raw file](https://github.com/robotlocomotion/drake/blob/9e57da35e16e10f3091ede342a07c32f43ae794d/drake/automotive/automotive_demo.cc#L109)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I agree, though no design for this exists yet. #5250 is a draft PR that uses Protocol Buffers to allow these types of parameters to be specified. Maybe we can discuss there?
</blockquote></details>

Thanks. I'm good with the design as-is. In general, I think the protobuf solution is probably the way to go. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-:-KhdyXHfpq_QrOzcTy78:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@soonho-tri for platform review please. Thanks!

---

Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-:-Khe3PXayBQV9NxwyqsB:b-lnsuqk)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 1 of 4 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 124 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-Khgr9yzfVEQwGjA-f2R:-Khgr9z-oPQNgYUsKoHl:boc4hjt) ([raw file](https://github.com/robotlocomotion/drake/blob/0317b405c0bc4c94e0e0ed5c272541cb1d13f6f1/drake/automotive/automotive_demo.cc#L124)):*
> ```cpp
>         lane_index * FLAGS_dragway_lane_speed_delta;
>     MaliputRailcarParams<double> params;
>     params.set_r(0);
> ```

BTW, `MaliputRailcarParams()` has `set_r(0)` and `set_h(0)`, which makes this and the following line redundant. When they are removed, you can make the above definition `const`.

---

*[drake/automotive/automotive_demo.cc, line 131 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-Khgry7mhyjOWQYAWpD6:-Khgry7mhyjOWQYAWpD7:bramcl8) ([raw file](https://github.com/robotlocomotion/drake/blob/0317b405c0bc4c94e0e0ed5c272541cb1d13f6f1/drake/automotive/automotive_demo.cc#L131)):*
> ```cpp
>     MaliputRailcarState<double> state;
>     const int row = i / FLAGS_num_dragway_lanes;
>     // 5 m ensures a gap between consecutive rows of Prius vehicles. It was
> ```

FYI, `5 m ensures a gap between consecutive rows of Prius vehicles...` should be moved to the up above where you define `kRailcarRowSpacing{5}`.

---

*[drake/automotive/automotive_demo.cc, line 133 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-KhgsVRgltwgsogUku8m:-KhgsVRgltwgsogUku8n:bhzsumu) ([raw file](https://github.com/robotlocomotion/drake/blob/0317b405c0bc4c94e0e0ed5c272541cb1d13f6f1/drake/automotive/automotive_demo.cc#L133)):*
> ```cpp
>     // 5 m ensures a gap between consecutive rows of Prius vehicles. It was
>     // empirically chosen.
>     state.set_s(initial_s_offset + kRailcarRowSpacing * row);
> ```

BTW, do we need any checking for the value that we use with `state.set_s()`? For example, if `initiali_s_offset` is too big, what happens?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5756)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 2 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 124 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-Khgr9yzfVEQwGjA-f2R:-KhhN2iBMtNfqc_Bbv0s:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0317b405c0bc4c94e0e0ed5c272541cb1d13f6f1/drake/automotive/automotive_demo.cc#L124)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, `MaliputRailcarParams()` has `set_r(0)` and `set_h(0)`, which makes this and the following line redundant. When they are removed, you can make the above definition `const`.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 131 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-Khgry7mhyjOWQYAWpD6:-KhhNQYsVND7e1zSUMKb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0317b405c0bc4c94e0e0ed5c272541cb1d13f6f1/drake/automotive/automotive_demo.cc#L131)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, `5 m ensures a gap between consecutive rows of Prius vehicles...` should be moved to the up above where you define `kRailcarRowSpacing{5}`.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 133 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-KhgsVRgltwgsogUku8m:-KhhO0GhDyHwAhPoboi8:bobczo0) ([raw file](https://github.com/robotlocomotion/drake/blob/0317b405c0bc4c94e0e0ed5c272541cb1d13f6f1/drake/automotive/automotive_demo.cc#L133)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, do we need any checking for the value that we use with `state.set_s()`? For example, if `initiali_s_offset` is too big, what happens?
</blockquote></details>

Done. I modified the code to throw an exception.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5756)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5756#-:-Khhfbb5tEp02R0xItca:bnfp4nl)*
<!-- Sent from Reviewable.io -->
